[action] add `xcodes` action, deprecating `xcversion` and `xcode-install`

rogerluan opened the pull request:
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I see several green `ci/circleci` builds in the "All checks have passed" section of my PR ([connect CircleCI to GitHub](https://support.circleci.com/hc/en-us/articles/360008097173-Why-aren-t-pull-requests-triggering-jobs-on-my-organization-) if not)
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

With the planned sunset of `xcode-install` (aka `xcversion`) gem (https://github.com/xcpretty/xcode-install/issues/467), a new action was needed to support `xcodes`.

### Description

This PR depends on https://github.com/RobotsAndPencils/xcodes/pull/220 and https://github.com/RobotsAndPencils/xcodes/pull/182 merge & release so the `installed` option with specified version is available, as well as `--install` and `--select` arguments for the `install` option.

These PRs were later merged & released under v1.1.0 https://github.com/RobotsAndPencils/xcodes/releases/tag/1.1.0, so v1.1.0 of xcodes is required to use and test this feature.

### Testing Steps

To test this branch, modify your Gemfile as:

```ruby
gem 'fastlane', git: 'https://github.com/fastlane/fastlane.git', branch: 'rogerluan-xcodes'
```

And run `bundle install` to apply the changes. 

Then switch your implementation of `xcversion` and `xcode_install` according to the migration guide which can be found here: https://github.com/xcpretty/xcode-install/blob/master/MIGRATION.md

Any and every testing is highly appreciated! I did very limited testing only 🙇 

<img width="734" alt="image" src="https://user-images.githubusercontent.com/8419048/191257574-7dd145a1-744a-48b0-956a-d1cb99a124c0.png">

<img width="604" alt="image" src="https://user-images.githubusercontent.com/8419048/199222112-c97e908b-2a5a-41c7-b414-6aa59ed36bbb.png">

rogerluan committed:
Move helpers to a separate and centralized file.

rogerluan committed:
Deprecate 'xcode-install' gem (the project has been sunset), adding 'xcodes' action to replace the 'xcode_install' one.

rogerluan committed:
Add 'select_only' option to 'xcodes', deprecating 'xcversion' action.

rogerluan committed:
Update documentation to state it also supports .xcode-version file.

rogerluan committed:
Add TODO for a future improvement.

rogerluan committed:
Lint.

rogerluan committed:
Fix link.

rogerluan committed:
Remove TODO comment.

rogerluan commented:
The spec failure that is happening in this PR also happens in master, and it may be something related to recent changes on Apple's end 😬

rogerluan commented:
It's weird:
- Locally, on my machine, a lot of specs fail both in this PR as well as in master, same for rubocop violations
- In CI, master seems to not be failing
- In CI, in another PR of mine, all specs passed
- In CI, this PR keeps failing in non-related changes 🤔

rogerluan committed:
Lint.

rogerluan committed:
Fix typo in error being raised.

rogerluan commented:
Ugh, so, locally, here's the breakdown:

- Ruby 3.1.x: a bunch of linter and specs issues
- Ruby 3.x.x: linter and specs are all passing
- Ruby 2.7.x: linter and specs are all passing
- Ruby 2.6.x: I'm having trouble installing this version using rbenv (M2 machine) 😥 so I couldn't test..

CI is failing specs and it is running Ruby 2.6.0 apparently... but not sure why this is happening in my PR only, and on presumably unrelated files (`fastlane_core/spec/cert_checker_spec.rb`) something related to:

```
#installed_wwdr_certificates
      should return installed certificate's alias (FAILED - 4)
      should return an empty array if unknown WWDR certificates are found (FAILED - 5)
```

😬 Any help would be appreciated

rogerluan committed:
Add ability to select Xcode just for current build, without sudo permissions.
- Depends on https://github.com/RobotsAndPencils/xcodes/pull/220

rogerluan committed:
Update documentation to point to 'xcodes' instead.

rogerluan committed:
Update xcodes link.

revolter commented on fastlane/lib/fastlane/actions/xcodes.rb:
```diff
@@ -0,0 +1,150 @@
+
+      def self.output
+        [
+          ['XCODES_XCODE_PATH', 'The path to the newly installed Xcode']
```
```suggestion
          ['XCODES_XCODE_PATH', 'The path to the newly installed Xcode version']
```

for consistency with the `return_value`.

revolter commented on fastlane/lib/fastlane/actions/xcodes.rb:
```diff
@@ -0,0 +1,150 @@
+        binary = params[:binary_path]
+        select_for_current_build_only = params[:select_for_current_build_only]
+        version = params[:version]
+        if params[:update_list] && !select_for_current_build_only
```
What would be the issue if we allow updating the list even if we want to select it for the current build only?

rogerluan replied:
Prior to release v1.0.0, updating the list required Apple authentication, so since update_list is true by default, if you just wanted to pass `select_for_current_build_only: true`, you wouldn't want to be bothered to authenticate with Apple. Now it doesn't require authentication anymore, so I don't see any issues with that!

@MattKiazyk could you confirm what is the exact behavior of the new auth-less requests in v1.0.0? Is it true that there are zero chances of requiring an authentication, regardless of the flow, regardless of which Xcode versions are already installed, macOS version, etc?

rogerluan replied:
If no user/pass is required EVER ☝️ then I can go ahead and remove all the user/pass management completely from this new action ✨

MattKiazyk replied:
@rogerluan  On Xcodes - if using the xcoderelease.com api (default) it is not required on the latest version. If using the Apple source it still would be.

rogerluan replied:
Thanks @MattKiazyk ! 🙏

rogerluan replied:
I just removed the requirement to provide xcodes action Apple user/pass 💪

revolter commented on fastlane/lib/fastlane/actions/xcodes.rb:
```diff
@@ -0,0 +1,150 @@
+
+        command = []
+        command << binary
+        command << "installed"
```
This doesn't work (anymore?):

```plaintext
[21:10:54]: --------------------
[21:10:54]: --- Step: xcodes ---
[21:10:54]: --------------------
[21:10:54]: $ /opt/homebrew/bin/xcodes installed '13.3'
[21:10:54]: ▸ Error: Unexpected argument '13.3'
[21:10:54]: ▸ Usage: xcodes installed [--directory <directory>] [--color] [--no-color]
[21:10:54]: ▸ See 'xcodes installed --help' for more information.
+------------------+------------------+
|            Lane Context             |
+------------------+------------------+
| DEFAULT_PLATFORM | ios              |
| PLATFORM_NAME    | ios              |
| LANE_NAME        | ios ensure_xcode |
+------------------+------------------+
[21:10:54]: Exit status of command '/opt/homebrew/bin/xcodes installed '13.3'' was 64 instead of 0.
Error: Unexpected argument '13.3'
Usage: xcodes installed [--directory <directory>] [--color] [--no-color]
  See 'xcodes installed --help' for more information.
```

I'm using `xcodes` version `1.0.0`.

rogerluan replied:
It doesn't work 'yet' 🙈 This PR this currently blocked by the release of the next version of `xcodes` 😬 as I stated in the PR description:

>This PR depends on https://github.com/RobotsAndPencils/xcodes/pull/220 merge & release so the installed option with specified version is available.

It has been merged, but not yet released. I've contacted the maintainer of xcodes (@MattKiazyk 🙌 ) and he said he doesn't have a release cadence, so yeah we need to wait for the next release to see this feature available

Any suggestions to enhance this @revolter ? 🙇

revolter replied:
> This PR depends on https://github.com/RobotsAndPencils/xcodes/pull/220 merge & release so the installed option with specified version is available.

Oh, I missed the "release" part 🤦🏻‍♂️ 😂

> Any suggestions to enhance this @revolter?

Enhance what?

rogerluan replied:
@revolter nevermind, I thought of adding a version check but there wouldn't be a great way to implement the same functionality before that feature gets released on `xcodes`. Let's just wait 🙇

rogerluan replied:
I ended up adding this version check (for v1.1.0) just for validation sake (it crashes if <1.1.0) so users are aware of this limitation 👍

revolter commented on fastlane/spec/helper/xcodes_helper_spec.rb:
```diff
@@ -0,0 +1,66 @@
+describe Fastlane::Helper::XcodesHelper do
+  describe "#read_xcode_version_file" do
```
This is more of a curiosity, but, what is the role of `#`?

rogerluan replied:
>use the Ruby documentation convention of . (or ::) when referring to a class method's name and # when referring to an instance method's name.

Source: https://www.betterspecs.org

(I learn these things as I go 😆 Ruby is not my mother tongue, clearly haha)

rogerluan replied:
🤔 Which made me think... `read_xcode_version_file` is actually a class method, not an instance one. I should probably fix this lol

rogerluan replied:
Done :)

revolter commented on fastlane/spec/helper/xcodes_helper_spec.rb:
```diff
@@ -0,0 +1,66 @@
+
+  describe "#find_xcodes_binary_path" do
+    it "returns the binary path when it's installed" do
+      expect(subject.class).to receive(:find_xcodes_binary_path).and_return("/opt/homebrew/bin/xcodes")
+      expect(subject.class.find_xcodes_binary_path).to eq("/opt/homebrew/bin/xcodes")
```
```suggestion
      expect(subject.class).to receive(:find_xcodes_binary_path).and_return("/path/to/bin/xcodes")
      expect(subject.class.find_xcodes_binary_path).to eq("/path/to/bin/xcodes")
```

for consistency with [xcodes_spec.rb](https://github.com/fastlane/fastlane/pull/20672/files#diff-2ca850ff6a7654f270c5deb87cbb737659ee2ee968e955924ae6524b170b78b5R6)?

rogerluan replied:
Oops, good catch! Left this behind by mistake after doing some local tests 😬

revolter commented on fastlane/spec/helper/xcodes_helper_spec.rb:
```diff
@@ -0,0 +1,66 @@
+      context "when argument is missing or empty" do
+        let(:argument) { nil }
+
+        it "raise user error" do
```
```suggestion
        it "raises user error" do
```

revolter commented on fastlane/spec/helper/xcodes_helper_spec.rb:
```diff
@@ -0,0 +1,66 @@
+        let(:argument) { nil }
+
+        it "raise user error" do
+          expect { subject.class::Verify.requirement(nil) }.to raise_error(FastlaneCore::Interface::FastlaneError, /Version must be specified/)
```
```suggestion
          expect { subject.class::Verify.requirement(argument) }.to raise_error(FastlaneCore::Interface::FastlaneError, /Version must be specified/)
```

revolter commented on fastlane/spec/helper/xcodes_helper_spec.rb:
```diff
@@ -0,0 +1,66 @@
+        let(:argument) { nil }
+
+        it "raise user error" do
+          expect { subject.class::Verify.requirement(nil) }.to raise_error(FastlaneCore::Interface::FastlaneError, /Version must be specified/)
```
```suggestion
          expect { subject.class::Verify.requirement(nil) }.to raise_error(FastlaneCore::Interface::FastlaneError, 'Version must be specified')
```

? Or am I missing something? 🤔

rogerluan replied:
Nope, you're right :)

rogerluan replied:
```suggestion
          expect { subject.class::Verify.requirement(argument) }.to raise_error(FastlaneCore::Interface::FastlaneError, 'Version must be specified')
```

revolter commented on fastlane/spec/helper/xcodes_helper_spec.rb:
```diff
@@ -0,0 +1,66 @@
+      context "when argument is not in RubyGems style" do
+        let(:argument) { "banana" }
+
+        it "raise user error" do
```
```suggestion
        it "raises user error" do
```

revolter commented on fastlane/spec/helper/xcodes_helper_spec.rb:
```diff
@@ -0,0 +1,66 @@
+        allow(Dir).to receive(:glob).with(".xcode-version").and_return([])
+      end
+
+      it "return nil" do
```
```suggestion
      it "returns nil" do
```

revolter reviewed (changes requested):
Hi @rogerluan 👋🏻

Nice to "see" you again after a long time 😃 And sorry for the late reply.

---

> 😬 Any help would be appreciated

I can't promise anything, but I hope I'll take a look some time after reviewing this.

---

I read the migration guide and this PR's description, I started testing it, but it failed with:

```plaintext
[21:00:46]: Couldn't find xcodes binary at path ''
[21:00:46]: You passed invalid parameters to 'xcodes'.
[21:00:46]: Check out the error below and available options by running `fastlane action xcodes`
+------------------+------------------+
|            Lane Context             |
+------------------+------------------+
| DEFAULT_PLATFORM | ios              |
| PLATFORM_NAME    | ios              |
| LANE_NAME        | ios ensure_xcode |
+------------------+------------------+
[21:00:46]: Invalid default value for binary_path, doesn't match verify_block
```

Although it is fairly obvious that the issue is that it can't find `xcodes`, and you have to (most probably) run `brew install robotsandpencils/made/xcodes`, I think that this should be mentioned in the migration guide and/or the new action's description, since it's a required dependency that's not automatically managed by _fastlane_. What do you think?

rogerluan committed:
Apply suggestions from code review.

Co-authored-by: Iulian Onofrei <5748627+revolter@users.noreply.github.com>

rogerluan committed:
Change "#" by "." in spec descriptions, to indicate class methods.

rogerluan committed:
Apply suggestion from code review.

rogerluan commented:
Oops, forgot to address the "no binary" issue. Gimme a few mins

rogerluan committed:
Improve documentation around requirements to run xcodes.

rogerluan commented:
Done @revolter 😊 Nice catch!

<img width="736" alt="image" src="https://user-images.githubusercontent.com/8419048/195832571-8ace11b0-d537-4c6c-bd52-5ef8a34a5e72.png">

rogerluan commented:
@revolter would you mind helping me understand what's failing in the CI specs? I added comments about it [here](https://github.com/fastlane/fastlane/pull/20672#issuecomment-1252236081) and I'm still clueless on how to solve this 😿

rogerluan committed:
Remove necessity to authenticate with Apple servers since xcodes v1.0.0.

rogerluan committed:
Use new --update and --selected arguments, which require xcodes v1.1.0.

revolter commented on fastlane/lib/fastlane/actions/xcodes.rb:
```diff
@@ -0,0 +1,135 @@
+
+        if (xcodes_args = params[:xcodes_args])
+          command << xcodes_args
+          shell_command = command.join(' ')
+          Actions.sh(shell_command)
```
Inline this `shell_command` variable, and also the one from line `36`?

revolter commented on fastlane/spec/actions_specs/xcodes_spec.rb:
```diff
@@ -0,0 +1,117 @@
+
+      describe "update_list argument" do
+        it "invokes update command when true" do
+          expect(Fastlane::Actions).to receive(:sh).with(/--update/)
```
```suggestion
          expect(Fastlane::Actions).to receive(:sh).with('--update')
```

rogerluan replied:
This is intentionally a regex, rspec treat string literals and regexes differently:

<img width="848" alt="image" src="https://user-images.githubusercontent.com/8419048/199477782-6223bba9-7511-4c7d-ba41-867dc5ce8b8a.png">

If I were to use a string, I'd not only have to use the entire exact string, but also expect the shell commands before and after the `/--update/` one 🤷 

This applies to all of the ones you commented on here

revolter replied:
True, my bad 🤦🏻‍♂️

revolter commented on fastlane/spec/actions_specs/xcodes_spec.rb:
```diff
@@ -0,0 +1,117 @@
+
+        it "doesn't invoke update command when false" do
+          expect(Fastlane::Actions).to receive(:sh).with("#{xcodes_binary_path} install '14' --select")
+          expect(Fastlane::Actions).to_not receive(:sh).with(/--update/)
```
```suggestion
          expect(Fastlane::Actions).to_not receive(:sh).with('--update')
```

rogerluan replied:
See above

revolter commented on fastlane/spec/actions_specs/xcodes_spec.rb:
```diff
@@ -0,0 +1,117 @@
+
+      it "doesn't invoke 'update' nor 'install' when select_for_current_build_only argument is true" do
+        expect(Fastlane::Actions).to_not receive(:sh).with(/--update|--install/)
+        expect(Fastlane::Actions).to receive(:sh).with(/installed/)
```
```suggestion
        expect(Fastlane::Actions).to receive(:sh).with('installed')
```

rogerluan replied:
See above

revolter reviewed (changes requested):
Just some small changes, to be less confusing for future contributors 🤔

revolter committed:
Fix comparing Gem::Version with String error

revolter committed:
Remove unused local variable

revolter committed:
Split long lines

revolter committed:
Fix lint issues

revolter commented:
@rogerluan, would you mind if I'd remove the merge commit, and rebase the branch instead?

revolter commented:
> Would that help debug the CI problem?

I hope it fixes it 😂 Because it doesn't make sense

revolter commented:
It doesn't :( I'll test some ideas this evening.

rogerluan committed:
Address style comments.

revolter committed:
Fix some WWDR certificates test failure

This was caused by the fact that that test calls a method that uses `$?`
(https://stackoverflow.com/a/6834572/865175), which, somehow, (I think)
was executed right after a call to XcodesHelper's
`find_xcodes_binary_path` method. This latter one returned a non-zero
code, because `xcodes` can't be installed on Ubuntu, which got picked up
by the certificates return code check, so it treated its actual command
as failed.

revolter committed:
Remove superfluous trailing slash

This makes it consistent with the rest of the codebase.

revolter committed:
Fix unsafe paths concatenation

revolter commented:
Ok, this took me a lot longer than I thought it would :| Very very weird issue, but great debugging nonetheless ❤️

I also noticed that some paths were being concatenated using `+`, so I used the same method as in `xcode_select.rb` (note: it works with an already present slash because that method makes sure not to duplicate the separator: https://stackoverflow.com/a/4114966/865175).

There are still some failures, but these ones might not be that hard to fix 🤔

_(Also, because I rebased and force-pushed to the branch, guessing that you didn't commit anything after 136b6648, please delete the local branch, fetch the new changes, and check it out again, to avoid other superfluous merges.)_

I really need to go to sleep now, as it's already tomorrow 😫

revolter commented:
Please let me work on this more? 😂 I have an idea that I want to test this evening.

revolter committed:
Replace `Actions.sh` call with a native one

(reverted from commit 11bd750dfdc65a548ffd210c77096edd153ee051)

revolter committed:
Fix error-prone shell command status check

In the case of the tests, `$?` was not mocked, so it looks like it was
indeed just a coincidence that it worked. Meaning, some other shell
command that somehow was actually executed (so it wasn't stubbed)
succeeded, which caused the `$?.success?` to succeed too.

revolter commented:
Ok, I was not happy with the fix, because I didn't understand it fully (still don't), and felt more like a hack. From what I (extensively) read, even through [the source code](https://github.com/ruby/open3/blob/master/lib/open3.rb), and by also testing it locally by replacing the `curl` command with `ls`, it looks like this (the last commit) is the safe and correct way to fix this issue for good.

revolter commented:
Somehow, the 2 failures (Validate Documentation and Validate Fastlane.swift generation) got fixed too 😮

rogerluan commented on fastlane_core/lib/fastlane_core/cert_checker.rb:
```diff
@@ -160,13 +160,13 @@ def self.install_wwdr_certificate(cert_alias)
       end
 
-      unless $?.success?
+      unless status.success?
```
👏👏👏

rogerluan commented on fastlane/lib/fastlane/helper/xcodes_helper.rb:
```diff
@@ -12,7 +12,7 @@ def self.read_xcode_version_file
 
       def self.find_xcodes_binary_path
-        `which xcodes`.strip
+        Actions.sh("which xcodes", log: false).strip
```
What the... 🤯

rogerluan commented:
@revolter dude, this was just amazing!! Thank you thank you thank you ❤️ Great context that you provided in the commit messages also! I'd never have guessed something like this, I can't even imagine how was your train of thought and how you got to the root cause 🤯 

Separately, do you think we should at least create an Issue for future improvement of getting rid of all the `$?` in other specs (and maybe in the source code too)?

Again, thank you so much 🙇 I'm so glad we'll be able to move this PR forward! 💪

rogerluan reviewed (commented):
@revolter dude, this was just amazing!! Thank you thank you thank you ❤️ Great context that you provided in the commit messages also! I'd never have guessed something like this, I can't even imagine how was your train of thought and how you got to the root cause 🤯 

Separately, do you think we should at least create an Issue for future improvement of getting rid of all the `$?` in other specs (and maybe in the source code too)?

Again, thank you so much 🙇 I'm so glad we'll be able to deliver this PR to `xcode-install` users  💪

revolter commented:
> I can't even imagine how was your train of thought and how you got to the root cause 🤯

Brute-force trial and error, using divide et impera, by reverting half of the changes, then half of the half, etc., until I pinpointed the responsible line of code, and then a lightbulb went off in my head about the `$?` usage from `cert_checker.rb`.

![image](https://user-images.githubusercontent.com/5748627/200051490-f6c6d4c5-235f-4fd2-ad9c-69c159c46180.png)

> Separately, do you think we should at least create an Issue for future improvement of getting rid of all the $? in other specs (and maybe in the source code too)?

Hmm, maybe that would be the way to go, but it's also a case of "if it ain't broken, don't fix it" 🤔 Maybe the best thing is that at least the 2 of us know of this, and hopefully remember it when (if) it happens in the future.

> Again, thank you so much 🙇 I'm so glad we'll be able to deliver this PR to `xcode-install` users  💪 

Really happy that I could (finally) help you with this one. Thank you too for the kind words, I'm really flattered ☺️ I'm happy that you took the time to do the other 99% of work for this migration ❤️

joshdholtz reviewed (approved):
This is soooo amazing!! ❤️ Sorry it took so long but I'm back from family leave and thank for doing this 😊
